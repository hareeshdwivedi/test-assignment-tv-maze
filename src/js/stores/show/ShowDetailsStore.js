import { observable, extendObservable, action } from 'mobx';
import Store from 'stores/Store';
import endpoints from 'transport/endpoints';

/**
 * The show's details store
 */
class ShowDetailsStore extends Store {

    /** The show's details */
    @observable
    showDetails = {};

    /** The show's episodes */
    @observable
    showEpisodes = [];

    /**
     * Get a show's details
     * @returns {Number} show's details
     */
    get ShowDetails() {
        return this.showDetails;
    }

    /**
     * Get a show's episodes
     * @returns {Number} show's episodes
     */
    get ShowEpisodes() {
        return this.showEpisodes;
    }

    /**
     * Set a show's details
     * @param {Object} showDetails - show's details to be set
     */
    @action
    setShowDetails = showDetails => {
        extendObservable(this.showDetails, showDetails);
    };

    /**
     * Set a show's episodes
     * @param {Object} showEpisodes - show's episodes to be set
     */
    @action
    setShowEpisodes = showEpisodes => {
        this.showEpisodes.replace(showEpisodes);
    };

    /**
     * Fetch a show's details
     * @param {Number} showId - a show identifier
     * @returns {Promise} a call api promise
     */
    fetchShowDetails(showId) {
        return this.Transport.callApi(endpoints.GET_SHOW_DETAILS, null, showId).then(this.setShowDetails);
    }

    /**
     * Fetch a show's episodes
     * @param {Number} showId - a show identifier
     * @returns {Promise} a call api promise
     */
    fetchShowEpisodes(showId) {
        return this.Transport.callApi(endpoints.GET_SHOW_EPISODES, null, showId).then(this.setShowEpisodes);
    }
}

export default ShowDetailsStore;
