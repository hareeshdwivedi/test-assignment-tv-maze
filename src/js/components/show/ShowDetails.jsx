import React, { Component } from 'react';
import { observer } from 'mobx-react';
import { Link } from 'react-router';
import PropTypes from 'prop-types';

import ShowDetailsStore from 'stores/show/ShowDetailsStore';

import * as styles from 'css/show-details.less';

/**
 * The component shows details of a show
 */
@observer
class ShowDetails extends Component {

    /** The component properties */
    static propTypes = {
        store: PropTypes.instanceOf(ShowDetailsStore)
    };

    /**
     * @see Component#componentWillMount()
     */
    componentWillMount() {
        const { store } = this.props;
        store.fetchShowDetails(1);
        store.fetchShowEpisodes(1);
    }

    /**
     * Render show's episodes
     * @param {Object} showEpisodes - show's episodes
     */
    renderEpisodes(showEpisodes) {
        return Object.entries(
            showEpisodes.reduce((acc, {name, season, number, image}) => {
                var episodes = acc[season];
                !episodes && (acc[season] = episodes = []);
                episodes.push({name, number, image});
                return acc;
            }, {})
        ).map(([season, episodes]) =>
            <div className="rtl-show-season" key={season}>
                <h2>Season {season}</h2>
                <ul>
                {
                    episodes.map(({name, image, number}) =>
                        <li key={`${season}_${number}`}>
                            <img src={image.medium} />
                            <Link to={`/show-episode/1/${season}/${number}`}>{name}</Link>
                        </li>)
                }
                </ul>
            </div>
        );
    }

    /**
     * @see Component#render()
     */
    render() {
        const { store: { ShowDetails, ShowEpisodes }} = this.props;
        if (Object.keys(ShowDetails).length) {
            const { name, image, summary} = ShowDetails;
            return (
                <div className="rtl-show">
                    <img src={image.original} />
                    <h1>{name}</h1>
                    <p dangerouslySetInnerHTML={{__html: summary}} />
                    <h2>Episodes</h2>
                    { this.renderEpisodes(ShowEpisodes) }
                </div>
            );
        }
        return null;
    }
}

export default ShowDetails;
