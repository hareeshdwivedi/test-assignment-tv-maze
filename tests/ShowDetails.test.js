import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import ShowDetailsStore from 'src/js/stores/show/ShowDetailsStore';
import ShowDetails from 'src/js/components/show/ShowDetails';

describe('Test ShowDetails component', () => {

    configure({adapter: new Adapter()});

    it('check initial markup of the component', () => {
        const mockedStore = new ShowDetailsStore();
        mockedStore.fetchShowDetails = jest.fn();
        mockedStore.fetchShowEpisodes = jest.fn();
        mockedStore.setShowDetails({name: 'name', image: {original: 'url'}, summary: 'summary'});
        mockedStore.setShowEpisodes([
            {name: 'name1', image: {medium: 'url1'}, season: 1, number: 1},
            {name: 'name2', image: {medium: 'url2'}, season: 1, number: 2}]);

        const renderedComponent = renderer.create(<ShowDetails store={mockedStore}/>);
        expect(renderedComponent.toJSON()).toMatchSnapshot();
    });
});