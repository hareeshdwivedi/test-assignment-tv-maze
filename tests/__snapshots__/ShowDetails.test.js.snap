// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ShowDetails component check initial markup of the component 1`] = `
<div
  className="rtl-show"
>
  <img
    src="url"
  />
  <h1>
    name
  </h1>
  <p
    dangerouslySetInnerHTML={
      Object {
        "__html": "summary",
      }
    }
  />
  <h2>
    Episodes
  </h2>
  <div
    className="rtl-show-season"
  >
    <h2>
      Season 
      1
    </h2>
    <ul>
      <li>
        <img
          src="url1"
        />
        <a
          onClick={[Function]}
          style={Object {}}
        >
          name1
        </a>
      </li>
      <li>
        <img
          src="url2"
        />
        <a
          onClick={[Function]}
          style={Object {}}
        >
          name2
        </a>
      </li>
    </ul>
  </div>
</div>
`;
